### DISCUSSION TITLE
best place for a tor patch

### DISCUSSION THREAD

**Discussion by jtracey** [shadow/shadow#3261] [General]
While continuing to poke around the performance of some Tor experiments, I noticed that the rounds were often getting bottlenecked on some Tor symbols. Turns out, they were symbols for some constant-time string operations. Since Shadow simulations have "infinite" computational speed, these shouldn't affect behavior versus their libc counterparts, but do seem to make a measurable difference for the simulation performance. I tried stubbing them out, and my (non-scientific) comparison showed about a 20% overall runtime improvement. This could be because of my particular configuration, but it at least makes some sense that these would be strict overhead, especially if it's in one heavy tor process blocking everything else. I know the openssl stubbing was dropped because it didn't matter much (and yeah, I didn't see any openssl symbols high in my measurements), so I'm not really sure the best place to put this now. It probably isn't necessary for most experiments, but it could be useful for some. Inline it in the docs on running Tor experiments? Put it as a patch file in tornettools somewhere?

Patch for reference:

```Diff
diff --git a/src/lib/ctime/di_ops.c b/src/lib/ctime/di_ops.c
index a6d69d4f91..8f8911bd9e 100644
--- a/src/lib/ctime/di_ops.c
+++ b/src/lib/ctime/di_ops.c
@@ -30,6 +30,7 @@
 int
 tor_memcmp(const void *a, const void *b, size_t len)
 {
+  return memcmp(a, b, len);
 #ifdef HAVE_TIMINGSAFE_MEMCMP
   return timingsafe_memcmp(a, b, len);
 #else
@@ -106,6 +107,7 @@ tor_memcmp(const void *a, const void *b, size_t len)
 int
 tor_memeq(const void *a, const void *b, size_t sz)
 {
+  return !memcmp(a, b, sz);
   /* Treat a and b as byte ranges. */
   const uint8_t *ba = a, *bb = b;
   uint32_t any_difference = 0;
```

**Comment by robgjansen** [comment]
I've generally kept Tor patches in my own fork of Tor at [robgjansen/tor](https://github.com/robgjansen/tor/branches). This change does seem a bit more specific to Shadow and less general than most of my research-based patches though.

I wonder if it might make sense to merge this directly into Tor? Presumably your patch would improve performance during regular Tor operation, but maybe less safely than they prefer? Would it make sense to propose it as a feature that it is only enabled when the `TestingTorNetwork` option is set?
